### PR TITLE
dm: MySQL 8.0 Compressed Binlogs

### LINKED ISSUE #6381: DM: Support MySQL 8.0
### Is your feature request related to a problem?

https://github.com/pingcap/tiflow/blob/77f8196f1b9a04a3f9cffd8bdaacffef64f5f566/dm/pkg/checker/mysql_server.go#L38-L48

When creating a task with a MySQL 8.0 source:
```json
			{
				"id": 0,
				"name": "mysql_version",
				"desc": "check whether mysql version is satisfied",
				"state": "warn",
				"errors": [
					{
						"severity": "warn",
						"short_error": "version suggested less than 8.0.0 but got 8.0.29"
					}
				],
				"extra": "address of db instance - 127.0.0.1:39135"
			}
```

### Describe the feature you'd like

Support MySQL 8.0

### Describe alternatives you've considered

_No response_

### Teachability, Documentation, Adoption, Migration Strategy

_No response_

### PR DESCRIPTION
**Pull request by dveeden**
### What problem does this PR solve?

Issue Number: close #6381

#### Tests 

 - Unit test
 - Integration test
 - Manual test (add detailed scripts or steps below)
 - No code

### Release note 

```release-note
Compatibility with MySQL 8.0 has been improved including reading compressed binlogs
```

### TODO

- Not assume `WRITE_ROWS_EVENTv2` (This is because `skipRowsEvent()` which is called from `handleRowsEvent` checks for this.
- Any good ways to reduce code duplication for handling events inside and outside of `TransactionPayloadEvent`?
- We now need to build with `make CGO=1 ...`. Is this ok? Do we need to update CI for this?
  - I've tried to fix this in 6df25dbf997960f6c1e6c0e10a1c2d7417f570b7 by always building with `CGO_ENABLED=1`

### CODE REVIEW

**Comment by lance6716** [comment]
great work! 
/cc @GMHDBJD

File: dm/pkg/checker/mysql_server.go [hunk 1]
```diff
@@ -35,14 +35,14 @@ func NewMySQLVersionChecker(db *sql.DB, dbinfo *dbutil.DBConfig) RealChecker {
 }
 
 // SupportedVersion defines the MySQL/MariaDB version that DM/syncer supports
-// * 5.6.0 <= MySQL Version < 8.0.0.
+// * 5.6.0 <= MySQL Version <= 8.0.24.
```
Comment (lance6716) [inline dm/pkg/checker/mysql_server.go:38]: we should also report error when found PARTIAL_UPDATE_ROWS_EVENT when increase the supported version

https://github.com/go-mysql-org/go-mysql/pull/774

**Review by lance6716 — COMMENTED** [review]
@dveeden It's not a draft PR now? we will start review soon

**Comment by GMHDBJD** [comment]
Can you add and test?

File: dm/pkg/checker/mysql_server.go (hunk 1 above)
Comment (lance6716) [reply dm/pkg/checker/mysql_server.go:38]: Hi, kindly remind that go-mysql has merged some commit about MySQL 8.0 features, please remember to update the library

Comment (dveeden) [reply dm/pkg/checker/mysql_server.go:38]: Done in d21bf42cb748f68de7721039ce09f9e45ef5c0af

File: dm/pkg/checker/mysql_server.go [hunk 2]
```diff
@@ -35,14 +35,14 @@ func NewMySQLVersionChecker(db *sql.DB, dbinfo *dbutil.DBConfig) RealChecker {
...
 var SupportedVersion = map[string]struct {
 	Min MySQLVersion
 	Max MySQLVersion
 }{
 	"mysql": {
 		MySQLVersion{5, 6, 0},
-		MySQLVersion{8, 0, 0},
+		MySQLVersion{8, 0, 24},
```
Comment (lance6716) [inline dm/pkg/checker/mysql_server.go:45]: I think we need more tests to officially support MySQL Version <= 8.0.24, let's make this change later

File: dm/syncer/syncer.go [hunk 3]
```diff
@@ -2344,6 +2344,35 @@ func (s *Syncer) Run(ctx context.Context) (err error) {
 			if err2 != nil {
 				return err2
 			}
+		case *replication.TransactionPayloadEvent:
+			for _, tpev := range ev.Events {
+				switch tpevt := tpev.Event.(type) {
+				case *replication.RowsEvent:
+					eventIndex++
+					s.metricsProxies.Metrics.BinlogEventRowHistogram.Observe(float64(len(tpevt.Rows)))
+					ec.header.EventType = replication.WRITE_ROWS_EVENTv2 // FIXME
```
Comment (lance6716) [inline dm/syncer/syncer.go:2353]: yes `handleRowsEvent` will use EventType to know it's a INSERT or DELETE or UPDATE. we can use the header of `tpevt`?

File: dm/syncer/syncer.go [hunk 4]
```diff
@@ -2344,6 +2344,35 @@ func (s *Syncer) Run(ctx context.Context) (err error) {
...
+					sourceTable, err2 = s.handleRowsEvent(tpevt, ec)
+					if sourceTable != nil && err2 == nil && s.cfg.EnableGTID {
+						if _, ok := affectedSourceTables[sourceTable.Schema]; !ok {
+							affectedSourceTables[sourceTable.Schema] = make(map[string]struct{})
+						}
+						affectedSourceTables[sourceTable.Schema][sourceTable.Name] = struct{}{}
+					}
+				case *replication.QueryEvent:
```
Comment (lance6716) [inline dm/syncer/syncer.go:2361]: seems `locationRecorder` should also handle `TransactionPayloadEvent` to correctly maintain the location property.

https://github.com/pingcap/tiflow/blob/d35ab225966b88cd6415b90a3d69b64dbcb65f54/dm/syncer/binlogstream/binlog_locations.go#L36-L55

File: dm/pkg/checker/mysql_server.go (hunk 2 above)
Comment (dveeden) [reply dm/pkg/checker/mysql_server.go]: Ok.  Updated in 6de0275701533b80fe6149e2e54d2173a1ac6a05

File: dm/syncer/syncer.go (hunk 3 above)
Comment (dveeden) [reply dm/syncer/syncer.go:2354]: Looks like it is not exported:
https://github.com/go-mysql-org/go-mysql/blob/f8546806eee3bcb9692350f670bd456d8bf1537c/replication/row_event.go#L833

```
> github.com/pingcap/tiflow/dm/syncer.(*Syncer).Run() ./dm/syncer/syncer.go:2352 (hits goroutine(543):1 total:1) (PC: 0x65d4ee4)
  2347:				}
  2348:			case *replication.TransactionPayloadEvent:
  2349:				for _, tpev := range ev.Events {
  2350:					switch tpevt := tpev.Event.(type) {
  2351:					case *replication.RowsEvent:
=>2352:						eventIndex++
  2353:						s.metricsProxies.Metrics.BinlogEventRowHistogram.Observe(float64(len(tpevt.Rows)))
  2354:						ec.header.EventType = replication.WRITE_ROWS_EVENTv2 // FIXME
  2355:						sourceTable, err2 = s.handleRowsEvent(tpevt, ec)
  2356:						if sourceTable != nil && err2 == nil && s.cfg.EnableGTID {
  2357:							if _, ok := affectedSourceTables[sourceTable.Schema]; !ok {
(dlv) p tpevt
*github.com/go-mysql-org/go-mysql/replication.RowsEvent {
	Version: 2,
	tableIDSize: 6,
	tables: map[uint64]*github.com/go-mysql-org/go-mysql/replication.TableMapEvent [
		124: *(*"github.com/go-mysql-org/go-mysql/replication.TableMapEvent")(0xc000eafc00), 
	],
	needBitmap2: false,
	eventType: WRITE_ROWS_EVENTv2 (30),
	Table: *github.com/go-mysql-org/go-mysql/replication.TableMapEvent {
		flavor: "",
		tableIDSize: 6,
		TableID: 124,
		Flags: 1,
		Schema: []uint8 len: 4, cap: 999901, [116,101,115,116],
		Table: []uint8 len: 2, cap: 999895, [116,49],
		ColumnCount: 1,
		ColumnType: []uint8 len: 1, cap: 999891, [3],
		ColumnMeta: []uint16 len: 1, cap: 1, [0],
		NullBitmap: []uint8 len: 1, cap: 999889, [0],
		SignednessBitmap: []uint8 len: 1, cap: 999886, [0],
		DefaultCharset: []uint64 len: 0, cap: 0, nil,
		ColumnCharset: []uint64 len: 0, cap: 0, nil,
		SetStrValue: [][][]uint8 len: 0, cap: 0, nil,
		setStrValueString: [][]string len: 0, cap: 0, nil,
		EnumStrValue: [][][]uint8 len: 0, cap: 0, nil,
		enumStrValueString: [][]string len: 0, cap: 0, nil,
		ColumnName: [][]uint8 len: 0, cap: 0, nil,
		columnNameString: []string len: 0, cap: 0, nil,
		GeometryType: []uint64 len: 0, cap: 0, nil,
		PrimaryKey: []uint64 len: 0, cap: 0, nil,
		PrimaryKeyPrefix: []uint64 len: 0, cap: 0, nil,
		EnumSetDefaultCharset: []uint64 len: 0, cap: 0, nil,
		EnumSetColumnCharset: []uint64 len: 0, cap: 0, nil,},
	TableID: 124,
	Flags: 1,
	ExtraData: []uint8 len: 0, cap: 999856, [],
	ColumnCount: 1,
	ColumnBitmap1: []uint8 len: 1, cap: 999855, [255],
	ColumnBitmap2: []uint8 len: 0, cap: 0, nil,
	Rows: [][]interface {} len: 1, cap: 1, [
		[
			*(*interface {})(0xc001fce4a0),
		],
	],
	SkippedColumns: [][]int len: 1, cap: 1, [
		[],
	],
	parseTime: false,
	timestampStringLocation: *time.Location nil,
	useDecimal: false,
	ignoreJSONDecodeErr: false,}
(dlv) p tpevt.eventType
WRITE_ROWS_EVENTv2 (30)
```

Should I create a PR for go-mysql to export it? or is there another way?

Comment (lance6716) [reply dm/syncer/syncer.go:2354]: Sorry I should say `tpev.Header.EventType`

File: dm/syncer/syncer.go (hunk 4 above)
Comment (dveeden) [reply dm/syncer/syncer.go:2362]: Something like this?
4e76bb961cc4e9b27f43d82277807055da9c7601

Or should it do something with the events embedded in the Transaction Payload Event?

File: dm/syncer/syncer.go (hunk 3 above)
Comment (dveeden) [reply dm/syncer/syncer.go:2354]:
```
(dlv) list
> github.com/pingcap/tiflow/dm/syncer.(*Syncer).Run() ./dm/syncer/syncer.go:2354 (hits goroutine(2074):1 total:1) (PC: 0x65d4f53)
  2349:				for _, tpev := range ev.Events {
  2350:					switch tpevt := tpev.Event.(type) {
  2351:					case *replication.RowsEvent:
  2352:						eventIndex++
  2353:						s.metricsProxies.Metrics.BinlogEventRowHistogram.Observe(float64(len(tpevt.Rows)))
=>2354:						ec.header.EventType = tpev.Header.EventType
  2355:						sourceTable, err2 = s.handleRowsEvent(tpevt, ec)
  2356:						if sourceTable != nil && err2 == nil && s.cfg.EnableGTID {
  2357:							if _, ok := affectedSourceTables[sourceTable.Schema]; !ok {
  2358:								affectedSourceTables[sourceTable.Schema] = make(map[string]struct{})
  2359:							}
(dlv) p tpev.Header.EventType
WRITE_ROWS_EVENTv2 (30)
```

Comment (dveeden) [reply dm/syncer/syncer.go]: 4554b518e4969fb7654ec2a172264403eabff8dd

File: dm/syncer/syncer.go (hunk 4 above)
Comment (dveeden) [reply dm/syncer/syncer.go:2362]: @lance6716 what do you think?

Comment (lance6716) [reply dm/syncer/syncer.go:2362]: I think https://github.com/pingcap/tiflow/commit/4e76bb961cc4e9b27f43d82277807055da9c7601 is good.

And sorry I don't have enough time to add integration test for it. we can do it later

**Review by GMHDBJD — APPROVED** [review]
LGTM

**Review by GMHDBJD — APPROVED** [review]
LGTM

**Pull request merged**